activestorage/test/models/attached/many_test.rb

```diff
@@ -578,6 +577,56 @@ def highlights
     end
   end
 
+  test "attaching a new blob from a Hash with a custom service" do
+    old_service = ActiveStorage::Blob.service
```

Comment:
You _could_ extract this temporary service setting behaviour into a private method that yields a block, like we've done with `append_on_assign` below.